imperial.js

```diff
@@ -3,23 +3,50 @@ const Action = require("./action");
 
 class Imperial {
   static fromLog(log) {
-    return new Imperial(log);
+    let game = new Imperial();
+    log.forEach((entry) => game.tick(entry));
+    return game;
   }
 
-  constructor(log) {
-    this.log = log;
+  constructor() {
+    this.log = [];
+    this.nations = this.setupNations();
+    this.players = {};
   }
 
   get state() {
     return {
       availableActions: this.availableActions(),
       investorCardHolder: this.investorCardHolder(),
-      nations: this.nations(),
-      players: this.players(),
+      nations: this.nationsState(),
+      players: this.playersState(),
       provinces: this.provinces(),
     };
   }
 
+  setupNations() {
+    let nations = {};
+    [Nation.AH, Nation.IT, Nation.FR, Nation.GB, Nation.GE, Nation.RU].forEach(
+      (nation) => (nations[nation] = { controller: "", treasury: 0 })
+    );
+    return nations;
+  }
```

Comment:
What is the output of this function? In the words of Inigo Montoya, "I do not think it means what you think it means." By way of abstract illustration, here's what I get when I do something similar with another Enum:

```es6
const { Enum } = require("./enum")
const Test = Enum.fromArray(["SPAM", "EGGS"])

const obj = [...Test].reduce((acc, e) => ({...acc, [e]: `got ${e.value}`}), {})
//=> {
//=>   '[object <anonymous>.SPAM]': 'got SPAM',
//=>   '[object <anonymous>.EGGS]': 'got EGGS'
//=> }
obj[Test.SPAM]
//=> "got SPAM"
obj[String(Test.SPAM)]
//=> "got SPAM"
Test.SPAM.value
//=> "SPAM"
obj[Test.SPAM.value]
//=> undefined
```

This code seems to work, although accidentally. As you may know, keys of objects are silently stringified in ECMAScript. `Enum` instances [implement stringification](https://github.com/Thrillberg/imperial/blob/54bd37439b77e7a73ca3392b6eb820f84856a7de/enum.js#L28-L30) as `${label}.${tag}`, which is what we see in the keys of our `obj`. (That extra `[object ...]` stuff comes from the ES interpreter).

I ran into this problem myself while first integrating the `Nation` enum, and I [solved it with a Map](https://github.com/Thrillberg/imperial/blob/54bd37439b77e7a73ca3392b6eb820f84856a7de/imperial.js#L352-L363). What do you think about converting this method to use a Map instead of an Object?

Reply:
Ah ok that almost makes sense. I think I'll screw this up a couple more times before I get it!